torch/onnx/experimental/fonnx/v1/exporter.py

```diff
@@ -0,0 +1,46 @@
+import io
+from typing import Optional
+
+import torch
+
+from .export_options import ExportOptions
+
+
+def dynamo_export(
+    torch_model: torch.nn.Module,
+    /,
+    *args,
+    options: Optional[ExportOptions] = None,
+    **kwargs,
+) -> ExportOutput:
+    """Export a torch.nn.Module to a ONNX graph
+
+    Args:
+        torch_model: User model to be exported
+        args: Positional inputs to `torch_model`
+        options: Advanced export options
+        kwargs: Keyword inputs to `torch_model`
+
+    Returns:
+        An in-memory representation of the exported ONNX model.
+        See ``ExportOutput`` for reference.
+    """
+    ...
+
+
+class ExportOutput:
+    """In-memory representation of a torch.nn.Module exported to ONNX
+
+    Args:
+        format: Format to save ONNX model."""
+
+    def __init__(self, format: onnx.ModelProto):
+        ...
+
+    def save(self, destination: str) -> None:
+        """Saves in-memory ONNX representation to `destination` using specified `format`"""
+        ...
+
+    def save_to_buffer(self) -> io.BytesIO:
+        """Saves ONNX representation into an in-memory bytes buffer"""
+        ...
```

Comment:
```python
def save(self, destination: Union[str, io.BytesIO]) -> None:
    pass
```

Reply:
I can't remember if we covered this, but we may want to stub `ExportOutputSerializer` to future-proof `save(...)`?

```python
from typing import Protocol

class ExportOutputSerializer(Protocol):
    def serialize(self, export_output: ExportOutput, destination: io.BytesIO) -> None:
        pass

class ProtobufExportOutputSerializer:
    def serialize(self, export_output: ExportOutput, destination: io.BytesIO) -> None:
        destination.write(export_output.onnx_model.SerializeToString())

class ExportOutput:
    def save(self, destination: Union[str, io.BytesIO], serializer: Optional[ExportOutputSerializer] = None):
        if serializer is None:
            serializer = ProtobufExportOutputSerializer()
        if isinstance(destination, str):
            with open(destination, "wb") as f:
                serializer.serialize(self, f)
        else:
            serializer.serialize(self, destination)
```

Reply:
Although I guess adding the optional `serializer` parameter in the future wouldn't actually constitute an API break in Python? I am a static, strongly typed languages guy 😄

Reply:
> ```python
> def save(self, destination: Union[str, io.BytesIO]) -> None:
>     pass
> ```
> 
> Returning `io.BytesIO` is odd. Wouldn't that imply that `save` is allocating _some kind_ of IO stream? If I know I want my own `io.BytesIO`, I will own it and call `save` to append the stream.

`save` returns None
`save_as_buffer` returns `io.BytesIO()`

Reply:
> Although I guess adding the optional `serializer` parameter in the future wouldn't actually constitute an API break in Python? I am a static, strongly typed languages guy 😄

Does it really make sense to write a super strict contract in something that at first will be internal only?

Reply:
done